### PR TITLE
Savegame refactor

### PR DESCRIPTION
**Pull request by rr-**
In preparation for #277 we need to clean up the existing savegame code which is, frankly, quite terrible in its current form.

This pull request:
- does the usual code convention refactors
- documents how savegames work
- simplifies the code (use the SG buffer for everything; remove level flags from `InitialiseLevel`; rename `SAVEGAME_INFO` which is horribly confusing to `GAME_INFO`)
- reduces file size of produced legacy saves from fixed size of ~ 10 KB down to actually used byte count.

It does not try to prepare for two savegame backends, that will be tackled later. I'd welcome some sanity checks if the changes so far did not break anything.

### CODE REVIEW

File: src/game/savegame.c [hunk 1]
```diff
@@ -493,229 +580,347 @@ void ExtractSaveGameInfo()
...
+    MYFILE *fp = File_Open(filename, FILE_OPEN_WRITE);
+    if (!fp) {
+        return false;
+    }
+    File_Write(&game_info->savegame_buffer[0], sizeof(char), m_SGBufPos, fp);
+    File_Close(fp);
+
+    REQUEST_INFO *req = &g_LoadSaveGameRequester;
+    req->item_flags[slot] &= ~RIF_BLOCKED;
+    sprintf(
```
Comment (walkawayy) [inline src/game/savegame.c:870]: What does this do?

File: src/global/types.h [hunk 2]
```diff
@@ -1449,36 +1449,28 @@ typedef struct START_INFO {
...
+        uint16_t all;
+        struct {
+            uint16_t available : 1;
+            uint16_t got_pistols : 1;
+            uint16_t got_magnums : 1;
+            uint16_t got_uzis : 1;
+            uint16_t got_shotgun : 1;
+            uint16_t costume : 1;
+        };
+    } flags;
```
Comment (walkawayy) [inline src/global/types.h:1462]: Why these particular variables of START_INFO to move into `flags`?

File: src/game/savegame.c (hunk 1 above)
Comment (rr-) [reply src/game/savegame.c:870]: After my changes it wrongfully updates the saved levels table in the passport with the level filename. Corrected to use level title instead :)

File: src/game/savegame.c [hunk 3]
```diff
@@ -210,61 +255,100 @@ void CreateStartInfo(int level_num)
...
+        0
+    };
+
+    SaveGame_WriteSG(&item_stats, sizeof(item_stats));
+
+    SaveGame_WriteSG(&g_FlipStatus, sizeof(int32_t));
     for (int i = 0; i < MAX_FLIP_MAPS; i++) {
-        int8_t flag = g_FlipMapTable[i] >> 8;
-        WriteSG(&flag, sizeof(int8_t));
+        int8_t flag = g_FlipMapTable[g_CurrentLevel] >> 8;
```
[the remark this answers is not in the document]
Comment (rr-) [reply src/game/savegame.c]: Yes. Good catch

**Comment by walkawayy** [comment]
I won't have time to look at this again and test until maybe tonight or probably tomorrow. I like the changes overall though!

**Comment by walkawayy** [comment]
Got a crash with this save: 
```
../src/game/savegame.c 391 SaveGame_ApplySaveBuffer Enabling Evil Lara savegame fix
```

[saveati-broken-evil-lara.zip](https://github.com/rr-/Tomb1Main/files/7874348/saveati-broken-evil-lara.zip)

**Comment by rr-** [comment]
Thanks for spotting this. I forgot to modify `SaveGame_NeedsEvilLaraFix` to include the new savegame buffer members. Fixed :)

**Comment by walkawayy** [comment]
I think it LGTM. Evil Lara saves work now.

**Review by walkawayy — APPROVED** [review]

**Pull request merged**